sys/net/netstats/netstats_neighbor.c

```diff
@@ -0,0 +1,400 @@
+         +  new_val * ewma_alpha) / NETSTATS_NB_EWMA_SCALE;
+}
+
+static void netstats_nb_update_etx(netstats_nb_t *stats, netstats_nb_result_t result,
+                                   uint8_t transmissions, bool fresh)
+{
+    /* don't do anything if driver does not report ETX */
+    if (transmissions == 0) {
+        return;
+    }
+
+    if (result != NETSTATS_NB_SUCCESS) {
+        transmissions = NETSTATS_NB_ETX_NOACK_PENALTY;
+    }
+
+#ifdef MODULE_NETSTATS_NEIGHBOR_ETX
+    stats->etx = _ewma(fresh, stats->etx, transmissions * NETSTATS_NB_ETX_DIVISOR);
+#else
+    (void)stats;
+    (void)result;
+    (void)transmissions;
+    (void)fresh;
+#endif
+}
+
+static void netstats_nb_update_time(netstats_nb_t *stats, netstats_nb_result_t result,
+                                    uint32_t duration, bool fresh)
+{
+    /* TX time already got a penalty due to retransmissions */
+    if (result != NETSTATS_NB_SUCCESS) {
+        duration *= 2;
+    }
+
+#if MODULE_NETSTATS_NEIGHBOR_TX_TIME
+    stats->time_tx_avg = _ewma(fresh, stats->time_tx_avg, duration);
+#else
+    (void)stats;
+    (void)result;
+    (void)duration;
+    (void)fresh;
+#endif
+}
+
+static void netstats_nb_update_rssi(netstats_nb_t *stats, uint8_t rssi, bool fresh)
+{
+#ifdef MODULE_NETSTATS_NEIGHBOR_RSSI
+    stats->rssi = _ewma(fresh, stats->rssi, rssi);
+#else
+    (void)stats;
+    (void)rssi;
+    (void)fresh;
+#endif
+}
+
+static void netstats_nb_update_lqi(netstats_nb_t *stats, uint8_t lqi, bool fresh)
+{
+#ifdef MODULE_NETSTATS_NEIGHBOR_LQI
+    stats->lqi = _ewma(fresh, stats->lqi, lqi);
+#else
+    (void)stats;
+    (void)lqi;
+    (void)fresh;
+#endif
+}
+
+static void netstats_nb_incr_count_tx(netstats_nb_t *stats, netstats_nb_result_t result)
+{
+#ifdef MODULE_NETSTATS_NEIGHBOR_COUNT
+    stats->tx_count++;
+
+    /* gracefully handle overflow */
+    if (stats->tx_count == 0) {
+        stats->tx_count = ~stats->tx_count;
+        stats->tx_count = stats->tx_count >> 4;
+        stats->tx_fail  = stats->tx_fail  >> 4;
+    }
+
+    if (result != NETSTATS_NB_SUCCESS) {
+        stats->tx_fail++;
+    }
+#else
+    (void)stats;
+    (void)result;
+#endif
+}
+
+static void netstats_nb_incr_count_rx(netstats_nb_t *stats)
+{
+#ifdef MODULE_NETSTATS_NEIGHBOR_COUNT
+    stats->rx_count++;
+#else
+    (void)stats;
+#endif
+}
+
+netstats_nb_t *netstats_nb_update_tx(netif_t *dev, netstats_nb_result_t result,
+                                     uint8_t transmissions)
+{
+    uint32_t now = xtimer_now_usec();
+    netstats_nb_t *stats;
+    uint32_t time_tx = 0;
+
+    _lock(dev);
+
+    /* Buggy drivers don't always generate TX done events.
+     * Discard old events to prevent the tx start <-> tx done correlation
+     * from getting out of sync. */
+    do {
+        stats = netstats_nb_get_recorded(dev, &time_tx);
+    } while (cib_avail(&dev->neighbors.stats_idx) &&
+            ((now - time_tx) > NETSTATS_NB_TX_TIMEOUT_MS * US_PER_MS));
```

Comment:
```suggestion
    } while (cib_avail(&dev->neighbors.stats_idx)
             && ((now - time_tx) > NETSTATS_NB_TX_TIMEOUT_MS * US_PER_MS));
```

Isn't this more readable?